RHDHBUGS-2625: Add procedure to configure Orchestrator to connect to existing PostgreSQL infrastructure (#2199)
redhat-developer/red-hat-developers-documentation-rhdh: modules/extend_orchestrator-in-rhdh/proc-configure-orchestrator-to-connect-to-existing-postgresql-infrastructure.adoc

:_mod-docs-content-type: PROCEDURE

[id="configure-orchestrator-to-connect-to-existing-postgresql-infrastructure_{context}"]
= Configure Orchestrator to connect to existing PostgreSQL infrastructure

[role="_abstract"]
Connect the Orchestrator plugins to your existing PostgreSQL database to leverage centralized database management and meet compliance requirements.

> By default, when you enable the Orchestrator plugin using the Operator, the `dependencies: - ref: sonataflow` field automatically provisions a `SonataFlowPlatform` custom resource (CR) and creates the required PostgreSQL database resources. The Operator uses specific naming patterns for these resources (such as `backstage-psql-{{backstage-name}}` for the Service and `backstage-psql-secret-{{backstage-name}}` for the Secret).

Reviewer: ```suggestion
By default, when you enable the Orchestrator plugin using the Operator, the `dependencies: - ref: sonataflow` field automatically provisions a `SonataFlowPlatform` custom resource (CR) and creates the required PostgreSQL database resources. The Operator uses specific naming patterns for these resources (such as `backstage-psql-{{backstage-name}}` for the service and `backstage-psql-secret-{{backstage-name}}` for the secret).
```

However, when you use an external PostgreSQL database that the Operator does not manage, these default resources and naming patterns do not exist. You must explicitly configure the `SonataFlowPlatform` CR to reference your external database resources and remove the automatic dependency provisioning.

.Prerequisites

> * You have installed {product-very-short} on {ocp-short}.

Reviewer: Is it really tied to OpenShift, or does it work on Kubernetes + the Operator framework?

Reviewer: I believe this would be more accurate. (Need SME approval):

```suggestion
* You have installed {product-very-short} by using the Operator.
```

Author: > Is it really tied to OpenShift, or does it work on Kubernetes + the Operator framework?

The procedure requires installing RHDH with the Operator, not a specific platform. The steps use the standard Kubernetes resources (Job, Service, Secret, ConfigMap) and the SonataFlowPlatform CR (Operator pattern). Hence the key requirement is the Operator installation method. I have updated the section.

> * You have {configuring-book-link}#configure-external-postgresql-databases_configuring-rhdh[configured an external PostgreSQL database] that is available and accessible from your {ocp-short} cluster, including a Kubernetes Service and Secret for the database connection.

Reviewer: Simplify, and avoid repeating implicit stuff.

```suggestion
* You have {configuring-book-link}#configure-external-postgresql-databases_configuring-rhdh[configured {product-short} to use an external PostgreSQL database].
```

> * You have access to create Jobs, Secrets, ConfigMaps, and custom resources in the namespace where you deploy the {product-custom-resource-type} CR.

Reviewer: ```suggestion
* You have access to create jobs, secrets, config maps, and custom resources in the namespace where you deploy the {product-custom-resource-type} CR.
```

.Procedure

> . Create the `backstage_plugin_orchestrator` database on your external PostgreSQL server by applying the following Job:

Reviewer: ```suggestion
. Create the `backstage_plugin_orchestrator` database on your external PostgreSQL server by applying the following job:
```

      persistence:
        postgresql:
          secretRef:
            name: __<SECRET-NAME-WITH-DB-CREDENTIALS>__
            userKey: POSTGRES_USER
            passwordKey: POSTGRES_PASSWORD
          serviceRef:
            name: __<SERVICE-NAME-TO-DB>__
            namespace: __<{product-very-short}-NAMESPACE>__
            databaseName: backstage_plugin_orchestrator
    jobService:
      enabled: true
      persistence:
        postgresql:
          secretRef:
            name: __<SECRET-NAME-WITH-DB-CREDENTIALS>__
            userKey: POSTGRES_USER
            passwordKey: POSTGRES_PASSWORD
          serviceRef:
            name: __<SERVICE-NAME-TO-DB>__
            namespace: __<{product-very-short}-NAMESPACE>__
            databaseName: backstage_plugin_orchestrator
----
+
[IMPORTANT]
====
Unlike the default configuration that uses the `dependencies: - ref: sonataflow` field to automatically provision database resources with specific naming patterns, this configuration explicitly references your external database Service and Secret. The `SonataFlowPlatform` CR will use these resources to connect to your external database instead of creating new database resources.
====

> . Configure the Orchestrator plugins in your dynamic plugins ConfigMap to remove the default `sonataflow` dependency and explicitly reference the SonataFlowPlatform services:

Reviewer: ```suggestion
. Configure the Orchestrator plugins in your dynamic plugins config map to remove the default `sonataflow` dependency and explicitly reference the SonataFlowPlatform services:
```
per naming convention, see https://redhat-documentation.github.io/supplementary-style-guide/#config-map

        disabled: false
      - package: "oci://registry.access.redhat.com/rhdh/red-hat-developer-hub-backstage-plugin-orchestrator-backend:{{inherit}}"
        disabled: false
        pluginConfig:
          orchestrator:
            dataIndexService:
              url: http://__<SERVICE-NAME-SONATAFLOW-PLATFORM-DATA-INDEX>__  # Typically sonataflow-platform-data-index-service
        dependencies: [{}]  # Empty array removes default 'ref: sonataflow' to prevent automatic database provisioning
      - package: "oci://registry.access.redhat.com/rhdh/red-hat-developer-hub-backstage-plugin-scaffolder-backend-module-orchestrator:{{inherit}}"
        disabled: false
        pluginConfig:
          orchestrator:
            dataIndexService:
              url: http://__<SERVICE-NAME-SONATAFLOW-PLATFORM-DATA-INDEX>__  # Typically sonataflow-platform-data-index-service
        dependencies: [{}]  # Empty array removes default 'ref: sonataflow' to prevent automatic database provisioning
      - package: "oci://registry.access.redhat.com/rhdh/red-hat-developer-hub-backstage-plugin-orchestrator-form-widgets:{{inherit}}"
        disabled: false
----

> . Update your {product-custom-resource-type} CR to reference the orchestrator plugin ConfigMap and inject the database credentials Secret:

Reviewer: ```suggestion
. Update your {product-custom-resource-type} CR to reference the orchestrator plugin config map and inject the database credentials Secret:
```

:_mod-docs-content-type: PROCEDURE

[id="configure-orchestrator-to-connect-to-existing-postgresql-infrastructure_{context}"]
= Configure Orchestrator to connect to existing PostgreSQL infrastructure

[role="_abstract"]

> Connect the Orchestrator plugins to your existing PostgreSQL database infrastructure to leverage centralized database management and meet compliance requirements. This configuration bypasses the default operator-managed database and integrates with external PostgreSQL services.

Reviewer: Low priority suggestion: I found "infrastructure" confusing in this context, and the second sentence repetitive / fluff.

```suggestion
Connect the Orchestrator plugins to your existing PostgreSQL database to leverage centralized database management and meet compliance requirements.
```

> By default, when you enable the Orchestrator plugin using the Operator, the `dependencies: - ref: sonataflow` field automatically provisions a `SonataFlowPlatform` CR and creates the required PostgreSQL database resources. The Operator uses specific naming patterns for these resources (such as `backstage-psql-{{backstage-name}}` for the Service and `backstage-psql-secret-{{backstage-name}}` for the Secret).

Reviewer: ```suggestion
By default, when you enable the Orchestrator plugin using the Operator, the `dependencies: - ref: sonataflow` field automatically provisions a `SonataFlowPlatform` custom resource (CR) and creates the required PostgreSQL database resources. The Operator uses specific naming patterns for these resources (such as `backstage-psql-{{backstage-name}}` for the Service and `backstage-psql-secret-{{backstage-name}}` for the Secret).
```

However, when you use an external PostgreSQL database that the Operator does not manage, these default resources and naming patterns do not exist. You must explicitly configure the `SonataFlowPlatform` CR to reference your external database resources and remove the automatic dependency provisioning.

.Prerequisites
* You have installed {product-very-short} on {ocp-short}.

> * You have an external PostgreSQL database available and accessible from your {ocp-short} cluster.

Reviewer: Consider adding a link to https://docs.redhat.com/en/documentation/red_hat_developer_hub/1.9/html/configuring_red_hat_developer_hub/configure-external-postgresql-databases_configuring-rhdh

* You have created a Kubernetes Service object in your namespace that points to the external PostgreSQL database. The cluster must resolve this Service from within.
+
[NOTE]
====
If your external database is outside the cluster, you can create a Service with an `ExternalName` type or an `Endpoints` object that maps to the external database IP address.

> ====

Reviewer: Consider adding this note rather to https://docs.redhat.com/en/documentation/red_hat_developer_hub/1.9/html/configuring_red_hat_developer_hub/configure-external-postgresql-databases_configuring-rhdh

* You have created a Kubernetes Secret containing the following database credentials:
** `POSTGRES_HOST`: The hostname or IP address of the external PostgreSQL server
** `POSTGRES_PORT`: The port number of the PostgreSQL service
** `POSTGRES_USER`: The database user name

> ** `POSTGRES_PASSWORD`: The database user password

Reviewer: Consider removing it as implicitly fullfilled when you have configured the external postgresl database (https://docs.redhat.com/en/documentation/red_hat_developer_hub/1.9/html/configuring_red_hat_developer_hub/configure-external-postgresql-databases_configuring-rhdh)

* You have access to create Jobs, Secrets, ConfigMaps, and custom resources in the namespace where you deploy the {product-custom-resource-type} CR.

.Procedure

> . Create the required Kubernetes resources for your external database connection. You can create the Service and `Secret` in parallel:

Reviewer: Consider moving this step 1 out and rather updating https://docs.redhat.com/en/documentation/red_hat_developer_hub/1.9/html/configuring_red_hat_developer_hub/configure-external- postgresql-databases_configuring-rhdh
where:
* Create a Kubernetes service is missing.
* Create a Kubernetes secret is already there.

(And this procedure should then focus on step 2 and following).

+
.. Create a Kubernetes Service that points to your external PostgreSQL database:
+
[source,yaml,subs="+attributes,+quotes"]
----
apiVersion: v1
kind: Service
metadata:
  name: external-postgresql-service  # Use this name in subsequent steps where <SERVICE-NAME-TO-DB> is referenced
spec:
  type: ExternalName  # Creates a CNAME record to the external database hostname

>   externalName: __<your-external-db-hostname>__  # FQDN of your external PostgreSQL server, for example, postgres.example.com

Reviewer: For consistency, move these comments out of the code blocks, into description lists following each code block.

  ports:
    - port: 5432
      targetPort: 5432
      protocol: TCP
----
+
[NOTE]
====
If your external database uses an IP address instead of a hostname, create a Service with Endpoints:

[source,yaml,subs="+attributes,+quotes"]
----
apiVersion: v1
kind: Service
metadata:
  name: external-postgresql-service
spec:
  ports:
    - port: 5432
      targetPort: 5432
      protocol: TCP
---
apiVersion: v1
kind: Endpoints
metadata:
  name: external-postgresql-service
subsets:
  - addresses:

>       - ip: __<your-external-db-ip>__  # IP address of your external PostgreSQL server

Reviewer: For consistency, move this comment out of the code block, into a description list.
